File: apps/api/src/lunaris_api/schemas/progress.py
from datetime import datetime

from pydantic import Field

from ..progress import LessonState
from .base import CamelModel


class ObjectiveProgressView(CamelModel):
    """One understood objective, keyed by its module + position (objectives carry no id)."""

    module_id: str
    objective_index: int
    understood_at: datetime


class LessonProgressView(CamelModel):
    """A lesson's learner state: in_progress on first open, done on completion."""

    lesson_id: str
    state: LessonState
    updated_at: datetime


class ProgressSummaryView(CamelModel):
    """Derived course-level rollup — recomputed per read, never stored."""

    understood_count: int
    objective_total: int
    lessons_done: int
    lesson_total: int
    percent: int


class ProgressSnapshotView(CamelModel):
    """The caller's progress on one course: the raw marks plus rollups derived against the
    course payload. ``summary`` is null (and ``kc_mastery`` empty) when the course itself isn't
    loadable — progress rows are independent of the payload."""

    course_id: str
    objectives: list[ObjectiveProgressView]
    lessons: list[LessonProgressView]
    summary: ProgressSummaryView | None = None
    kc_mastery: dict[str, bool] = Field(default_factory=dict)


class ObjectiveMarkRequest(CamelModel):
    """Mark (or un-mark) one module objective as understood. Bounds mirror the DB checks."""

    module_id: str = Field(min_length=1, max_length=200)
    objective_index: int = Field(ge=0, le=999)
    understood: bool


class LessonMarkRequest(CamelModel):
    """Advance a lesson's learner state (idempotent upsert)."""

    lesson_id: str = Field(min_length=1, max_length=200)
    state: LessonState

File: apps/web/src/lib/progress.ts
import { authedFetch } from "./apiClient";

/** A lesson's learner state — in_progress on first open, done on completion. */
export type LessonState = "in_progress" | "done";

/** One understood objective, keyed by its module + position (objectives carry no id). */
export interface ObjectiveProgress {
  moduleId: string;
  objectiveIndex: number;
  understoodAt: string;
}

export interface LessonProgress {
  lessonId: string;
  state: LessonState;
  updatedAt: string;
}

/** Derived course-level rollup — recomputed by the API per read, never stored. */
export interface ProgressSummary {
  understoodCount: number;
  objectiveTotal: number;
  lessonsDone: number;
  lessonTotal: number;
  percent: number;
}

/** The caller's progress on one course — raw marks plus rollups derived against the course
 *  payload (summary null / kcMastery empty when the course isn't loadable). `lastOpenedAt` /
 *  `lastLessonId` are null until the learner first opens the course. */
export interface CourseProgress {
  courseId: string;
  objectives: ObjectiveProgress[];
  lessons: LessonProgress[];
  summary?: ProgressSummary | null;
  kcMastery?: Record<string, boolean>;
  lastOpenedAt?: string | null;
  lastLessonId?: string | null;
}

export async function fetchCourseProgress(
  apiBaseUrl: string,
  courseId: string,
  signal?: AbortSignal,
): Promise<CourseProgress> {
  const response = await authedFetch(
    `${apiBaseUrl}/api/courses/${encodeURIComponent(courseId)}/progress`,
    { signal: signal ?? null },
  );
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  return (await response.json()) as CourseProgress;
}

async function putProgress(
  apiBaseUrl: string,
  courseId: string,
  path: string,
  mark: unknown,
): Promise<void> {
  const response = await authedFetch(
    `${apiBaseUrl}/api/courses/${encodeURIComponent(courseId)}/progress/${path}`,
    { method: "PUT", headers: { "Content-Type": "application/json" }, body: JSON.stringify(mark) },
  );
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
}

export function putObjectiveProgress(
  apiBaseUrl: string,
  courseId: string,
  mark: { moduleId: string; objectiveIndex: number; understood: boolean },
): Promise<void> {
  return putProgress(apiBaseUrl, courseId, "objective", mark);
}

export function putLessonProgress(
  apiBaseUrl: string,
  courseId: string,
  mark: { lessonId: string; state: LessonState },
): Promise<void> {
  return putProgress(apiBaseUrl, courseId, "lesson", mark);
}

/** Record that the learner opened this course — at a lesson when given, else a bare touch
 *  (the server preserves any previously recorded position). */
export function putCourseOpened(
  apiBaseUrl: string,
  courseId: string,
  lastLessonId?: string,
): Promise<void> {
  return putProgress(apiBaseUrl, courseId, "opened", lastLessonId ? { lastLessonId } : {});
}
